src/lib/annotations/doc/DocDrawingThread.js

```diff
@@ -0,0 +1,88 @@
+import {
+    DRAW_POINTER_UP,
+    DRAW_POINTER_DOWN,
+    PAGE_PADDING_TOP,
+    PAGE_PADDING_BOTTOM,
+    CLASS_ANNOTATION_LAYER_DRAW
+} from '../annotationConstants';
+import { getScale } from '../annotatorUtil';
+import DrawingPath from '../drawing/DrawingPath';
+import DrawingThread from '../drawing/DrawingThread';
+import * as DocAnnotatorUtil from './docAnnotatorUtil';
+
+class DocDrawingThread extends DrawingThread {
+    /** @property {number} - Drawing state */
+    lastPage;
+
+    /** @property {HTMLElement} - Page element being observed */
+    pageEl;
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+    /**
+     * Handle a pointer movement
+     *
+     * @param {Object} location The location information of the pointer
+     * @return {void}
+     */
+    handleMove(location) {
+        const pageChanged = this.lastPage && this.lastPage !== location.page;
+
+        this.lastPage = location.page;
+        if (!this.pageEl || pageChanged) {
+            this.pageEl = DocAnnotatorUtil.getPageEl(this.annotatedElement, location.page);
+            if (pageChanged) {
+                this.handleStop(location);
+            }
+            return;
+        }
+
+        const [x, y] = DocAnnotatorUtil.getBrowserCoordinatesFromLocation(location, this.pageEl);
+        if (this.drawingFlag === DRAW_POINTER_DOWN) {
+            this.pendingPath.addCoordinate(x, y);
+        }
+    }
+
+    /**
+     * Start a drawing stroke
+     *
+     * @return {void}
+     */
+    handleStart() {
+        this.drawingFlag = DRAW_POINTER_DOWN;
+        const scale = getScale(this.annotatedElement);
+        const context = DocAnnotatorUtil.getContext(
+            this.pageEl,
+            CLASS_ANNOTATION_LAYER_DRAW,
+            PAGE_PADDING_TOP,
+            PAGE_PADDING_BOTTOM
+        );
+        const canvas = context.canvas;
+
+        this.savedState = context.getImageData(0, 0, canvas.width, canvas.height);
```

Comment:
Each render clears the canvas, puts the savedState onto the canvas, then draws the path of all of the current point. It seems that drawImage(canvas) is faster than get/putImageData ( as a result, it would be good to have a memory canvas) but I am unsure how it performs in contrast to redrawing all of the paths on each render.